src/vg-tooltip.ts

```diff
@@ -0,0 +1,850 @@
+    if (shouldShowTooltip(item)) {
+      // clear existing promise because mouse can only point at one thing at a time
+      cancelPromise();
+
+      // make a new promise with time delay for tooltip
+      tooltipPromise = window.setTimeout(function () {
+        init(event, item, options);
+      }, options.delay || DELAY);
+    }
+  });
+
+  // update tooltip position on mouse move
+  // (important for large marks e.g. bars)
+  vgView.on("mousemove.tooltipUpdate", function (event: MouseEvent, item: SceneGraph) {
+    if (shouldShowTooltip(item) && tooltipActive) {
+      update(event, item, options);
+    }
+  });
+
+  // clear tooltip on mouse out
+  vgView.on("mouseout.tooltipClear", function (event: MouseEvent, item: SceneGraph) {
+    if (shouldShowTooltip(item)) {
+      cancelPromise();
+
+      if (tooltipActive) {
+        clear(event, item, options);
+      }
+    }
+  });
+
+  return {
+    destroy: function () {
+      // remove event listeners
+      vgView.off("mouseover.tooltipInit");
+      vgView.off("mousemove.tooltipUpdate");
+      vgView.off("mouseout.tooltipClear");
+
+      cancelPromise(); // clear tooltip promise
+    }
+  } 
+};
+
+/* Cancel tooltip promise */
+function cancelPromise() {
+  /* We don't check if tooltipPromise is valid because passing
+   an invalid ID to clearTimeout does not have any effect
+   (and doesn't throw an exception). */
+  window.clearTimeout(tooltipPromise);
+  tooltipPromise = undefined;
+}
+
+/* d3mapping from fieldDef.type to formatType */
+var formatTypeMap: {[type: string]: 'number' | 'time'} = {
+  "quantitative": "number",
+  "temporal": "time",
+  "ordinal": undefined,
+  "nominal": undefined
+}
+
+/**
+* (Vega-Lite only) Supplement options with vlSpec
+*
+* @param options - user-provided options
+* @param vlSpec - vega-lite spec
+* @return the vlSpec-supplemented options object
+*
+* if options.showAllFields is true or undefined, vlSpec will supplement
+* options.fields with all fields in the spec
+* if options.showAllFields is false, vlSpec will only supplement existing fields
+* in options.fields
+*/
+function supplementOptions(options: Option, vlSpec: Spec) {
+  // fields to be supplemented by vlSpec
+  var supplementedFields: Field[];
+
+  // if showAllFields is true or undefined, supplement all fields in vlSpec
+  if (options.showAllFields !== false) {
+    vl.spec.fieldDefs(vlSpec).forEach(function (fieldDef) {
+      // get a fieldOption in options that matches the fieldDef
+      var fieldOption = getFieldOption(options.fields, fieldDef);
+
+      // supplement the fieldOption with fieldDef and config
+      var supplementedFieldOption = supplementFieldOption(fieldOption, fieldDef, vlSpec);
+
+      supplementedFields.push(supplementedFieldOption);
+    });
+  }
+  // if showAllFields is false, only supplement existing fields in options.fields
+  else {
+    if (options.fields && options.fields.length > 0) {
```

Comment:
No need to check `options.fields.length > 0`   (if your length is zero, forEach won't run)